Unit tests

petervdonovan opened the pull request:
This PR includes changes to the structure of reactor-c. The intention is to make unit testing easier.

For this to work, it appears that we will indeed have to build `reactor-c` as a library. Fortunately, this is probably not very hard -- we just have to parameterize the build process according to whether the execution is threaded, federated, federated and decentralized, etc. To my knowledge, it is an open question whether the library is destined to be used for production, or if it is only going to be used for testing.

lhstrh commented on core/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+endmacro()
+
+# Applies the X as a preprocessor definition if X is defined.
+macro(define X)
```
What is this for?

petervdonovan replied:
The user passes arguments to CMake via the command line, like
```
cmake .. -DNUMBER_OF_WORKERS=4 -DFEDERATED=1
```

This macro passes those variables along to the compiler as preprocessor definitions. The changes I just added to the README will try to explain this.

lhstrh commented on test/general/data-structures/vector_test.c:
```diff
@@ -0,0 +1,139 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include "core/utils/vector.h"
```
Maybe the directory of this test file should mirror the path of the code under test (i.e., `utils` instead of `data-structures`)?

petervdonovan replied:
Sounds good -- the change has been applied.

lhstrh commented on test/general/data-structures/vector_test.c:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * @brief Runs a randomly selected test on `v`.
```
```suggestion
 * @brief Run a randomly selected test on `v`.
```

lhstrh commented on test/general/data-structures/vector_test.c:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * @brief Checks that voting does not cause an error.
```
```suggestion
 * @brief Check that voting does not cause an error.
```

lhstrh commented on test/general/data-structures/vector_test.c:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * @brief Checks that the result of a random access to `v` yields the
```
```suggestion
 * @brief Check that the result of a random access to `v` yields the
```

lhstrh commented on test/general/data-structures/vector_test.c:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * @brief Tests the "push all" functionality of `v`.
```
```suggestion
 * @brief Test the "push all" functionality of `v`.
```

lhstrh commented on test/general/data-structures/vector_test.c:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * @brief Tests the pop functionality of `v`.
```
```suggestion
 * @brief Test the pop functionality of `v`.
```

lhstrh commented on test/general/data-structures/vector_test.c:
```diff
@@ -0,0 +1,139 @@
+static int distribution[4] = {30, 50, 5, 15};
+
+/**
+ * @brief Tests the append functionality of `v`.
```
```suggestion
 * @brief Test the append functionality of `v`.
```

lhstrh commented on test/Tests.cmake:
```diff
@@ -0,0 +1,38 @@
+# Add the appropriate directories for the provided build parameters.
+add_test_dir(${TEST_DIR}/general)
+if(NUMBER_OF_WORKERS)
+    add_test_dir(${TEST_DIR}/threaded)
```
I personally like `single-threaded` and `multi-threaded` more because `unthreaded` is a bit of a misnomer...

petervdonovan replied:
Now that you mention it, I see that you are right, at least in principle.

The code often includes the words "unthreaded," "non-threaded," and "threaded." Examples from grep:
```
core/reactor.h: * This file contains header information used by both the threaded and
core/reactor.h: * non-threaded versions of the C runtime. A generated C program will have
core/trace.h: * @param worker The thread number of the worker thread or 0 for unthreaded execution.
core/reactor.c:/* Runtime infrastructure for the non-threaded version of the C target of Lingua Franca. */
core/reactor.c:            tracepoint_reaction_starts(reaction, 0); // 0 indicates unthreaded.
core/reactor_threaded.c: * A struct representing a barrier in threaded 
core/reactor_common.c: * This is the internal implementation shared by both the threaded
core/reactor_common.c: * and non-threaded versions.
```
We could open an issue for this terminology. In principle, it seems like a problem, but there could be reasons for it. For example,
* The "unthreaded" runtime doesn't use a threading library.
* It must be distinguished from the multithreaded runtime in the case where `NUMBER_OF_WORKERS` equals 1.
* "unthreaded" and "threaded" have fewer syllables than "single-threaded" and "multithreaded."

petervdonovan replied:
Well, this change has been applied -- no more misnomer. Since our use of the word "unthreaded" in other places is a separate issue, it seems okay to mark this one as resolved.

lhstrh commented on README.md:
```diff
@@ -15,3 +15,17 @@ To generate and view documentation, follow the following steps:
+To create a new test, write a C program with a file name ending in "test.c"
+in the `test` directory.
+
+To run unthreaded tests, execute the following. Note that -U is required to undefine
```
There are certain tests that we'll want to have for both the multi-threaded and single-threaded runtime. Would we just have to replicate these tests? Or might there be a more elegant solution for this?

petervdonovan replied:
Currently, the tests under "general" are executed always, regardless of whether the runtime is single-threaded or multi-threaded. Does that suffice?

If tests become too expensive to run, then we could make an additional category for tests that can be run separately.

lhstrh replied:
Not sure... There are differing implementations (reactor.c vs reactor_threaded.c) of various functions that should functionally behave the same and thus would require verbatim copies of certain tests, with the difference only being the imports. This seems like a situation where the Cmake potentially could help.

petervdonovan replied:
I think that use case should be covered if those tests are placed in "general." In that case, you could run `cmake .. -UNUMBER_OF_WORKERS` to run the tests in "general", compiling and linking `.c` files from the single-threaded runtime, and you could run `cmake .. -DNUMBER_OF_WORKERS=2` (or something else nonzero) if you want to run the same tests, compiling and linking `.c` files from the multithreaded runtime.

It works on my machine, at least.

lhstrh replied:
OK, yes, that sounds good. Thanks!

lhstrh reviewed (approved):
This looks great, @petervdonovan! Should this still be a draft? It looks to me like we could merge this... I left a few comments/questions.

petervdonovan commented:
> Should this still be a draft? It looks to me like we could merge this... 

Well, we seem to have a couple of options, in order of increasing complexity:
1. Merge in what we have now, so that it the build is configurable from the command line according to whether it is single-threaded or multi-threaded
2. Add more changes so that the build is configurable according to whether it is federated; add federated tests
3. Make the build fully configurable, so that it can replace a big chunk of the logic that currently exists in the code generator, including some of `CCmakeGenerator.xtend` as well as the parts that determine which files to include.

I lean toward addressing the comments and cleaning up one thing, and then making items 2 and 3 a separate PR.

lhstrh commented:
> > Should this still be a draft? It looks to me like we could merge this...
> 
> Well, we seem to have a couple of options, in order of increasing complexity:
> 
>     1. Merge in what we have now, so that it the build is configurable from the command line according to whether it is single-threaded or multi-threaded
> 
>     2. Add more changes so that the build is configurable according to whether it is federated; add federated tests
> 
>     3. Make the build fully configurable, so that it can replace a big chunk of the logic that currently exists in the code generator, including some of `CCmakeGenerator.xtend` as well as the parts that determine which files to include.
> 
> 
> I lean toward addressing the comments and cleaning up one thing, and then making items 2 and 3 a separate PR.

I agree with you! Please proceed as proposed :-)
